pkg/querier/tenantfederation/merge_queryable.go

```diff
@@ -184,22 +234,29 @@ func (m *mergeQuerier) Close() error {
 // tenantLabelName.
 func (m *mergeQuerier) Select(sortSeries bool, hints *storage.SelectHints, matchers ...*labels.Matcher) storage.SeriesSet {
 	matchedTenants, filteredMatchers := filterValuesByMatchers(defaultTenantLabel, m.tenantIDs, matchers...)
-	var seriesSets = make([]storage.SeriesSet, 0, len(matchedTenants))
-	for pos, tenantID := range m.tenantIDs {
-		if _, matched := matchedTenants[tenantID]; !matched {
+	var seriesSets = make([]storage.SeriesSet, len(matchedTenants))
+	var ssPos int
+	var wg sync.WaitGroup
+	for tenantPos := range m.tenantIDs {
```

Comment:
This code might be a good candidate for using [`concurrency.ForEachUser`](https://github.com/cortexproject/cortex/blob/master/pkg/util/concurrency/runner.go#L16)?

Reply:
After this merges, I plan a PR which will generalise the mergeQueryable, so it can be used along other dimensions that just "users". So probably it's not the best fit, but I think you made me aware that we probably want to be able to limit the maximum concurrency and for that [concurrency.ForEach](https://github.com/cortexproject/cortex/blob/d302dc418e8a2c8ec87b600319e815e44d465106/pkg/util/concurrency/runner.go#L67) might be a good fit.

Reply:
That makes sense - using `ForEach` should be fine then.

Reply:
Thinking about after a bit. I wonder if others think too high concurrency is a problem that should be addressed at this stage? It feels now a bit too early for that from my point of view...

Reply:
After having caught up with @jtlisi outside of this issue. I have implemented a limit using concurrency.ForEach and a constant value of `16`.

Reply:
Nice! Looks good. I think it was worth doing regardless (subjective).